### ISSUE TITLE
Chat online member list for users

### ISSUE THREAD

**Issue by cameronurnes** [owncast/owncast#3344]
My users frequently ask if there is a way to view who else is in the chat room.  Something similar to the user list available in the admin panel, but available in the normal website/chat box.

**Comment by gabek** [comment]
This is not something that is currently available.

**Comment by cameronurnes** [comment]
Yep, this is meant as a feature request not a bug report.  Thanks for your work!

**Comment by gabek** [comment]
This is mostly not available on purpose. If somebody wants to quietly lurk and not be known that they're there, I think that's valid. If somebody wants others to know they're there, they'll say something in chat.

I understand that in some communities, knowing who is there at all times might be what's wanted, but I haven't heard much of a request for this kind of thing. Usually I hear the opposite, they want more ways to hide who's there (removing join messages, hide viewer counts, etc).

**Comment by jprjr** [comment]
I'll second that having the viewer list available would be great!

Usually when I want to lurk and not be known I'm there - I use a private window, so I generate a new, random user. In my experience people don't usually pay attention to new users with a default username until they say something.

As far as settings go it could probably be part of hiding join/leave messages. Figure if you don't want join/leave messages, you probably aren't interested in showing who's there, either.

**Comment by gabek** [comment]
I'll add this to the backlog, since I understand why it could be wanted. But in general, this does seem like the opposite of what most people ask for.

Just to add some additional context, any feature request that requires new UI or some amount of design is less likely to be acted upon in the short term because that's where this project is lacking. Sure, it could be hacked together, and some random button could be added that pops up a list of names, but I don't really want to add things like this without some amount of thought around the user experience, and we don't have a dedicated design resource.

**Comment by cameronurnes** [comment]
I still get asked about this semi-regularly when I make my group use owncast for a night instead of kosmi.  Hopefully it can make it in eventually!

**Comment by gabek** [comment]
One item of confusion, that will automatically arise if this were to go in, is people would get very confused when they see 10 viewers but only 7 people listed in chat. They'd either think Owncast is broken, or somebody is watching twice to try to increase the viewer count, or something else incorrect. I know it's not necessarily my problem if people don't understand, but if they think things are broken because now they have more information than they did before, then it does kind of become a problem.

**Comment by cameronurnes** [comment]
My view count is already a bit wrong anyways since I'm using an external monitor to make sure the stream is OK and auto-restart it if it fails.  I definitely won't be complaining about that issue, at least.  Would be great if this chat list feature ended up as an optional checkbox not turned on by default or something.

**Comment by gabek** [comment]
It's not the streamers I'm concerned about, they should know how things work. It's the viewers I'm concerned about.

**Comment by germainelee** [comment]
Since the lurkers who wants to remain anonymous, what are the chances of them typing in the chat? Let's say they want to chat online and remain anonymous, maybe we can provide the option to chat anonymously. And only the streamer would be able to see his profile.

I have a suggestion that could protect their privacy at the same time: 
We keep the view count as it is, and have a viewer list which shows the streamer's name and a list of viewers who are active "chatters". In other words, the viewer's name doesn't show in the chat until he types something. Since it would be a dynamic list of viewers who are active "chatters", viewer's names would only go away after 5 minutes if they don't type anything. But we don't show the count of active Chatters to avoid any confusion.

**Comment by gabek** [comment]
Hi there, thanks for chiming in to discuss this!

The anonymous chat user is the primary concern I had as well, and I don't think there's a perfect way to balance that. It's kind of up to the owner of the Owncast server. If the owner has join and part messages enabled, then it will broadcast that a new person joined the chat. And the suggestion that we enable the user list if join/part messages are enabled feels like a good compromise. But that's not up to the user, that's up to the admin. It's not a perfect solution, but I think it's fair.

I like your idea of keeping somebody hidden until they type something. But unfortunately, that would break the existing join/part message functionality. Since if we left it how it is, it would tell everyone the moment you joined, defeating the point of not showing you in the list, and if we didn't show the join message, then it's a feature that couldn't be trusted, since it would only work after somebody said something. And once they said something, you know they're there anyway, not really needing the join message.

I'm totally with you for not showing a total count of people in the chat, as that would be really confusing when it's compared to the viewer count. I think that's already going to be really confusing when people count the people in the list and compare it to the viewer count and think something is wrong.

I guess one question to ask is how common do we think it is that somebody really wants to hide? And because you can use a completely anonymous chat name, is that good enough for somebody to be anonymous even though technically people can see they've joined chat?

**Comment by jprjr** [comment]
Idea - right now when a stream has chat enabled, viewing the stream and joining the chat is all one single "thing." You pull up the stream, you get assigned a username if you don't have one, you join the chat whether you intended to chat or not, and the join message goes out (assuming that's enabled).

What if this was split up so when a user pulls up the stream, the chat appears in a view-only mode with a "Click here to start chatting" type button where the text box is. Then on clicking the button - you get assigned a username (or use your existing token to connect), the join message goes out, and so on.

This would allow lurkers to watch without announcing their presence, if they so choose, because now chatting is an opt-in feature.

I suspect there would need to be some retooling to how chat works now. I believe connecting to the chat websocket requires a user token as a query parameter, you'd probably need to have the initial connection be anonymous, then perform user lookup by sending the token over the socket. But this could all be tied to the "click here to start chatting" button.

**Comment by gabek** [comment]
That could be a solution. But I'm still curious what number of people are we optimizing for here? Do we want to make chat more complex for 99% of people (including ourselves) because of a hypothetical 1%? Maybe. I know these people are out there. But they're far from a majority. And the more I think about it, the more I think the real solution is this one person per server to ask the admin "hey, I don't feel comfortable, can you turn off the chat list?". And they either do or they don't, it's their server.

I think we'd be more likely to have problems with adding a roadblock in front of chat, not to mention the investment in web work that would be required. 

I'm also 100% up for not adding the chat list at all. I don't think it's necessary personally, and it does feel it has more downsides than positives if we believe that some percentage of people are going to be made personally uncomfortable by it without us entirely changing how chat works. The less frontend chat work that needs to be done, the better! :)

**Comment by germainelee** [comment]
I understand that there might be a limited community to optimise the chat feature. I believe there are multiple scenarios a host and/or his community would may or may not want the participant list made public in the livestream of their event, so correct me if any of the scenarios is not possible –

- when I'm having a Private watch party with family, showing active participant list to Viewers is a great way of fostering a sense of community and engagement.
- when I'm Broadcasting gaming session to public, Only admin can view the complete list of members so that people feel comfortable watching the stream while engaging in the chat. By doing so, this reinforces viewer privacy and preventing misuse.

To cater to a range of use cases, while balancing the need for interaction with the need for privacy, I agree that only hosts should have the control to enable/disable chat list. This could be a toggle option to show/hide participant list view for viewers in the admin settings. So I'm thinking there's a invite-only/private viewing type of livestream would have this option.

Would probably need more research... before putting in development effort and barely anybody uses it. Even I don't check the participant list on YouTube live, Twitch, etc. :X

**Comment by cameronurnes** [comment]
Just to pass on the message, a couple weeks ago one of my users said: "If there was a user list and we could send images, we could ditch kosmi entirely"  (That's the other site they mostly use)

I'd definitely be happy with just a checkbox for the server owner to enable/disable the user list!  Pretty sure they could do without the images (but would be nice if that was a checkbox option for the server owner too).

The idea of having opt-in chat seems like a separate feature than this.

**Comment by germainelee** [comment]
@cameronurnes  would you mind sharing your opinion on the mindset of why the users would like to view a user list? This will help me understand what's needed to balance community engagement and prioritize privacy. 

I can see the benefits of having viewer list available to audience, but without this understanding it would be challenging for me to proceed.

**Comment by cameronurnes** [comment]
@germainelee  Its a community group where we all know each other, meet up IRL at events around the country a few times per year, and even do a cruise.  Most people in the group have been around for years.  People like to see who is there when they join the stream/watch party, since they are familiar names not strangers.  Like looking around the room when you arrive at a party.  Owncast already does join messages, so it's not like a user won't see everyone who joins after them, this just lets them see who is already there.

I use owncast and dizquetv to push video 24/7 for people who just want to hang out even if the nightly host is gone from kosmi.  Kosmi video also looks terrible and frequently has issues, so using owncast which provides great video quality would be preferable, but so far most hosts don't want to swap yet.  This is cited as one of the reasons.

There is of course the occasional anonymous user who is not logged in, and that feature would still protect the privacy of those who don't want to share their name that night.  However in our case, most users are logged in with a consistent username over time, and even the anon users are almost certainly in the FB group.  It's definitely not a massive number of viewers, but we have over 50 during our yearly "big" stream, about 20 regulars on every Saturday night, and about 10 on most weeknights.  There's about 1k people in the FB group around it, tho most only really drop in once in awhile.

We rarely have any of the issues or concerns a publicly posted stream might have with a lot of strangers around.
